File: pdl/infra/kotlin/src/main/kotlin/io/pdl/infrastructure/data/QueryBuilder.kt
package io.pdl.infrastructure.data

class QueryBuilder private constructor(
    private val table: String,
    private var store: DBStore?
) {
    private val filters = mutableListOf<Filter>()
    private val projection = mutableListOf<String>()
    private val orderings = mutableListOf<OrderClause>()
    private var limit: Int? = null
    private var offset: Int? = null

    fun withStore(store: DBStore?): QueryBuilder {
        this.store = store
        return this
    }

    fun filter(column: String, operator: Operator, value: Any?): QueryBuilder {
        filters += Filter(column, operator, value)
        return this
    }

    fun project(vararg columns: String): QueryBuilder {
        projection += columns
        return this
    }

    fun offset(value: Int): QueryBuilder {
        offset = value
        return this
    }

    fun limit(value: Int): QueryBuilder {
        limit = value
        return this
    }

    fun range(offset: Int, limit: Int): QueryBuilder {
        this.offset = offset
        this.limit = limit
        return this
    }

    fun asc(column: String): QueryBuilder = orderBy(column, OrderDirection.ASC)

    fun desc(column: String): QueryBuilder = orderBy(column, OrderDirection.DESC)

    fun orderBy(column: String, direction: OrderDirection): QueryBuilder {
        orderings += OrderClause(column, direction)
        return this
    }

    fun load(): List<Map<String, Any?>> {
        val resolved = StoreRegistry.resolve(store)
        return resolved.select(table, filters, projection, orderings, limit, offset)
    }

    fun delete(primaryKey: String) {
        val resolved = StoreRegistry.resolve(store)
        val rows = resolved.select(table, filters, listOf(primaryKey), emptyList(), null, null)
        rows.forEach { resolved.delete(table, primaryKey, it) }
    }

    companion object {
        @JvmStatic
        fun table(name: String, store: DBStore? = null): QueryBuilder = QueryBuilder(name, store)
    }
}

File: pdl/infra/php/src/Io/Pdl/Infrastructure/Data/QueryBuilder.php
<?php

namespace Io\Pdl\Infrastructure\Data;

class QueryBuilder
{
    private array $filters = [];
    private array $projection = [];
    private array $orderings = [];
    private ?int $limit = null;
    private ?int $offset = null;

    public function __construct(
        private string $table,
        private ?DBStore $store
    ) {
    }

    public function withStore(?DBStore $store): self
    {
        $this->store = $store;
        return $this;
    }

    public function filter(string $column, string $operator, mixed $value): self
    {
        $this->filters[] = [$column, $operator, $value];
        return $this;
    }

    public function project(string ...$columns): self
    {
        $this->projection = array_merge($this->projection, $columns);
        return $this;
    }

    public function offset(int $value): self
    {
        $this->offset = $value;
        return $this;
    }

    public function limit(int $value): self
    {
        $this->limit = $value;
        return $this;
    }

    public function range(int $offset, int $limit): self
    {
        $this->offset = $offset;
        $this->limit = $limit;
        return $this;
    }

    public function orderBy(string $column, string $direction): self
    {
        $this->orderings[] = [$column, $direction];
        return $this;
    }

    public function load(): array
    {
        $store = StoreRegistry::resolve($this->store);
        return $store->select(
            $this->table,
            $this->filters,
            $this->projection,
            $this->orderings,
            $this->limit,
            $this->offset
        );
    }

    public function delete(string $primaryKey): void
    {
        $store = StoreRegistry::resolve($this->store);
        $rows = $store->select(
            $this->table,
            $this->filters,
            [$primaryKey],
            [],
            null,
            null
        );
        foreach ($rows as $row) {
            $store->delete($this->table, $primaryKey, $row);
        }
    }
}
